### PR TITLE
Fix the drawText hook

### PR DESCRIPTION
**Pull request by checkraisefold**
A problem we've recently been facing is, because of MSVC's lack of naked function support on the x64 architecture, the prologue generated by the hook was overwriting the stack pointer and wiping the varargs away. This means we couldn't simply call the original game function and have it work if Austin was using the varargs at all for formatting. The pretty much only workable solution is below.

Using some revolutionary assembly that took me a few hours to make that I didn't bother commenting, mainly because MASM is really ass. This is ready to merge, but I need you to review and test on your machine to make sure it doesn't just work on my device. I believe this has some caveats, but with the current calls to drawText Alex does, there should be no problems. This fixes problems with undefined behavior where the varargs weren't being handled properly.
Explanation of the ASM:

pushVarArgs is passed the address of the final argument before the varargs in the hook, and the amount of arguments. ref: https://blog.aaronballman.com/2012/06/how-variable-argument-lists-work-in-c/

It then loops through the arguments given in the hook and pushes them onto the stack after offset 0x48, because offset 0x48 is where the "alpha" argument will be on the stack when calling the original game function.
```
pushVarArgs proc
	readloop:
		mov r8d, dword ptr [rcx + rdx * 8h]
		mov [rsp + 48h + rdx * 8h], r8d
		sub rdx, 1
	jnz readloop
	ret
pushVarArgs endp
```

Because pushVarArgs leaves the stack with information after 0x48, we will get shit tons of runtime errors, stack corruption, and undefined behavior if we don't clear the stack. This is why we have the `clearStack` proc.

The clearStack proc is given the same argument count as the pushVarArgs function, then writes 0x0000 to the values on the stack we previously wrote to. This completely solves our problem of corrupting the stack.
```
clearStack proc
	readloop:
		mov dword ptr [rsp + 48h + rcx * 8h], 0000
		sub rcx, 1
	jnz readloop
	ret
clearStack endp
```

i spent 4.5 hours on this help me god

### CODE REVIEW

**Comment by windows-fryer** [comment]
Wow awesome! I don't care though.

**Comment by noche-x** [comment]
seems pretty solid, i'll test this to see whats going on linux.

**Comment by checkraisefold** [comment]
I've added Linux support (UNTESTED. I AM TOO LAZY TO SET UP A TEST ENVIRONMENT, PLEASE TEST IT @noche-x). The only caveat is the end user compiling with MSVC on Windows has to install the VSNASM build customization found at https://github.com/ShiftMediaProject/VSNASM
However, this build customization is well worth it. It is very well made and its only caveat is not being able to step through ASM made with it in the VS debugger, but you can just set a breakpoint in Cheat Engine or x64dbg and step through that way. It gives you detailed error and warning information in the error list, unlike MASM where I have to go digging through the output and looking for the MASM compiler exe's output.

**Comment by noche-x** [comment]
Going to test this before I PR it.

**Comment by noche-x** [comment]
With some adjustments it doesn't work..
pic-window-211220-1521-43
.

File: CMakeLists.txt [hunk 1]
```diff
@@ -12,10 +12,16 @@ set(CMAKE_CXX_STANDARD 20)
...
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -fno-rtti")
 
+enable_language(ASM_NASM)
+
 if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} --no-gnu-unique")
 endif()
 
+if(UNIX)
+	string(APPEND CMAKE_ASM_NASM_FLAGS "-dUNIX=1")
```
Comment (noche-x) [inline CMakeLists.txt:22]: Causes a compiling error, can be fixed by removing Threads::Threads.

File: src/game.hpp [hunk 2]
```diff
@@ -20,7 +20,7 @@ class game
 
     std::uintptr_t drawText;
     #ifdef _WIN32
-    std::add_pointer_t<int64_t (char *, float, float, float, int, float, float, float, float, void*)> drawTextFunc;
+    std::add_pointer_t<int64_t (char *, float, float, float, int, float, float, float, float, ...)> drawTextFunc;
     #else
     std::add_pointer_t<int64_t (char *, int, int, int, float, float, float, float, float, float, float, int)> drawTextFunc;
```
Comment (noche-x) [inline src/game.hpp:25]: Linux function signature missing var args.

File: src/hooks.cpp [hunk 3]
```diff
@@ -20,6 +20,14 @@
 #include <cmath>
 #include <ctime>
 
+#ifdef _WIN32
+extern "C" void __fastcall pushVarArgs(void* addr, long long count);
+extern "C" void __fastcall clearStack(long long count);
```
Comment (noche-x) [inline src/hooks.cpp:25]: Use FASTCALL macro instead of ifdef __win32

File: src/hooks.cpp [hunk 4]
```diff
@@ -46,17 +54,31 @@ int64_t drawHud(int64_t arg1)
...
+	if (argCount > 0)
+		clearStack(static_cast<long long>(argCount));
 #else
-	auto ret = g_game->drawTextFunc(text, params | TEXT_SHADOW, a, b, x, y, scale, red, green, blue, alpha, c);
+	std::string_view textStr = text;
+	auto argCount = std::count(textStr.begin(), textStr.end(), '/'); // this will break if alex uses a / escape or some shit (lol)
+	if (argCount > 0) {
+		pushVarArgs(&alpha, static_cast<long long>(argCount));
+	}
+	auto ret = g_game->drawTextFunc(text, params | TEXT_SHADOW, x, y, scale, red, green, blue, alpha);
```
Comment (noche-x) [inline src/hooks.cpp:79]: Only put this and the Windows original function call in ifdefs, because everything else is the same.

**Review by noche-x — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (checkraisefold) [reply CMakeLists.txt:22]: I can't find the source of the threads package or whatever so I don't know why this happens

**Comment by checkraisefold** [comment]
ready to merge, also fixes msvc compilation
@noche-x for re-review

**Comment by checkraisefold** [comment]
refuses to inject under windows now? looking into it. suspecting something broke with logger

**Comment by checkraisefold** [comment]
fixed everything except the main menu text.. the credits menu works, the main menu doesn't render the sub rosa custom text. not sure what changed

**Review by noche-x — APPROVED** [review]
Works correctly on Linux, compiles, everything seems fine.

**Pull request merged**